src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,189 @@
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        T tmp = current.value;
+        current.value = value;
+        return tmp;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = findNodeByIndex(index);
+        unlink(current);
+        return current.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        int index = findIndexByValue(object);
+        if (index == -1) {
+            return false;
+        } else {
+            unlink(findNodeByIndex(index));
+        }
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Incorrect index " + index);
+        }
+    }
+
+    private void checkIndexForAddMethod(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Incorrect index " + index);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        if (index == 0) {
+            return head;
+        }
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current;
+    }
+
+    private int findIndexByValue(T object) {
+        Node<T> current = head;
+        int index = 0;
+        while (current != null) {
+            if (current.value == object || current.value != null && current.value.equals(object)) {
+                return index;
+            } else {
+                current = current.next;
+                index++;
+            }
+        }
+        return -1;
+    }
+
+    private void unlink(Node<T> nodeToRemove) {
+        if (nodeToRemove == head) {
+            if (size == 1) {
+                head = null;
+                tail = null;
+            } else {
+                head = head.next;
+                head.prev = null;
+            }
+            size--;
+        } else {
+            Node<T> current = head;
+            while (current != null) {
+                if (current == nodeToRemove) {
+                    if (current == tail) {
+                        tail = tail.prev;
+                        tail.next = null;
+                    } else {
+                        current.prev.next = current.next;
+                        current.next.prev = current.prev;
+                    }
+                    size--;
+                    return;
+                }
+                current = current.next;
+            }
+        }
```

Comment:
```suggestion
        final T value = nodeToRemove.value;
        if (node.prev == null) {
            head = node.next;
        } else {
            node.prev.next = node.next;
        }

        if (node.next == null) {
            tail = node.prev;
        } else {
            node.next.prev = node.prev;
        }

        size--;
        return value;
    }
```
let's simplify this a bit

Reply:
you killed me)